test_conformance/generic_address_space/atomic_tests.cpp

```diff
@@ -0,0 +1,231 @@
+// limitations under the License.
+
+#include <vector>
+
+#ifdef __APPLE__
+#include <OpenCL/opencl.h>
+#else
+#include <CL/cl.h>
+#endif
+
+#include "errorHelpers.h"
+#include "typeWrappers.h"
+
+namespace {
+// In this source, each workgroup will generate one value.
+// Every other workgroup will use either a global or local
+// pointer on an atomic operation.
+const char* KernelSourceInvariant = R"OpenCLC(
+kernel void testKernel(global int* globalPtr, local int* localPtr) {
+    int gid = get_global_id(0);
+    int tid = get_local_id(0);
+    int wgid = get_group_id(0);
+    int wgsize = get_local_size(0);
+
+    if (tid == 0) *localPtr = 0;
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    // Initialise the generic pointer to
+    // the global.
+    generic int* ptr = globalPtr + wgid;
+
+    // In a workgroup-invariant way, select a localPtr instead.
+    if ((wgid % 2) == 0)
+        ptr = localPtr;
+
+    int inc = atomic_fetch_add(ptr, 1);
+
+    // In the cases where the local memory ptr was used,
+    // save off the final value.
+    if ((wgid % 2) == 0 && inc == (wgsize-1))
+        globalPtr[wgid] = inc;
+}
+)OpenCLC";
+
+// In this source, each workgroup will generate two values.
+// Every other work item in the workgroup will select either
+// a local or global memory pointer and perform an atomic
+// operation on that.
+const char* KernelSourceVariant = R"OpenCLC(
+kernel void testKernel(global int* globalPtr, local int* localPtr) {
+    int gid = get_global_id(0);
+    int tid = get_local_id(0);
+    int wgid = get_group_id(0);
+    int wgsize = get_local_size(0);
+
+    if (tid == 0) *localPtr = 0;
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    // Initialise the generic pointer to
+    // the global.  Two values are written per WG.
+    generic int* ptr = globalPtr + (wgid * 2);
+
+    // In a workgroup-invariant way, select a localPtr instead.
+    if ((tid % 2) == 0)
+        ptr = localPtr;
+
+    int inc = atomic_fetch_add(ptr, 1);
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    // In the cases where the local memory ptr was used,
+    // save off the final value.
+    if ((tid % 2) == 0)
+        globalPtr[(wgid * 2) + 1] = inc;
+}
+)OpenCLC";
+}
+
+int test_generic_atomics_invariant(cl_device_id deviceID, cl_context context,
+                                   cl_command_queue queue, int num_elements)
+{
+    const auto version = get_device_cl_version(deviceID);
+
+    if (version < Version(2, 0)) return TEST_SKIPPED_ITSELF;
+
+    cl_int err = CL_SUCCESS;
+    clProgramWrapper program = clCreateProgramWithSource(
+        context, 1, &KernelSourceInvariant, nullptr, &err);
+    test_error(err, "clCreateProgramWithSource failed");
+
+    err = clBuildProgram(program, 1, &deviceID, "-cl-std=CL2.0", nullptr,
+                         nullptr);
+    test_error(err, "clBuildProgram failed");
+
+    clKernelWrapper kernel = clCreateKernel(program, "testKernel", &err);
+    test_error(err, "clCreateKernel failed");
```

Comment:
It would be best to use `create_single_kernel_helper` here.  Among other things, it will handle passing the right `-cl-std` version, which is important because many OpenCL 3.0 devices do not support OpenCL C 2.0.

Reply:
Should be done now